File: oxidize-golang/internal/generate/cache.go
package generate

import (
	"fmt"
	"sync"

	"github.com/Zapdev-labs/oxidize/golang/core/model"
)

// CachedModel holds a loaded model and tokenizer path for reuse across requests.
type CachedModel struct {
	Path     string
	Result   LoaderResult
	Warning  string
	Backend  string
	FellBack bool
}

// ModelCache stores GGUF models keyed by path + loader settings.
type ModelCache struct {
	mu    sync.RWMutex
	items map[string]*CachedModel
}

// DefaultModelCache is the process-wide server cache.
var DefaultModelCache = &ModelCache{items: make(map[string]*CachedModel)}

func (c *ModelCache) cacheKey(path string, cfg LoaderConfig) string {
	return fmt.Sprintf("%s|%s|%d|%d|%d|%s", path, cfg.Backend, cfg.ContextSize, cfg.NGPULayers, cfg.GPUs, cfg.Parallelism)
}

// Get returns a cached entry if present.
func (c *ModelCache) Get(path string, cfg LoaderConfig) (*CachedModel, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	item, ok := c.items[c.cacheKey(path, cfg)]
	return item, ok
}

// Load returns a cached model or loads and stores it.
func (c *ModelCache) Load(path string, cfg LoaderConfig) (CachedModel, error) {
	key := c.cacheKey(path, cfg)
	c.mu.RLock()
	if item, ok := c.items[key]; ok {
		c.mu.RUnlock()
		return *item, nil
	}
	c.mu.RUnlock()

	result, err := LoadModelFromPath(path, cfg)
	if err != nil {
		return CachedModel{}, err
	}
	entry := &CachedModel{
		Path:     path,
		Result:   result,
		Warning:  result.Warning,
		Backend:  result.Backend,
		FellBack: result.FellBack,
	}
	c.mu.Lock()
	c.items[key] = entry
	c.mu.Unlock()
	return *entry, nil
}

// InferenceFromCache loads and returns *InferenceModel when possible.
func InferenceFromCache(path string, cfg LoaderConfig) (*model.InferenceModel, CachedModel, error) {
	entry, err := DefaultModelCache.Load(path, cfg)
	if err != nil {
		return nil, CachedModel{}, err
	}
	inference, ok := entry.Result.Model.(*model.InferenceModel)
	if !ok {
		return nil, entry, fmt.Errorf("generate: expected InferenceModel, got %T", entry.Result.Model)
	}
	return inference, entry, nil
}

File: oxidize-python/oxidize_python/internal/generate/cache.py
"""Process-wide GGUF model cache mirroring oxidize-golang/internal/generate/cache.go."""

from __future__ import annotations

import threading
from dataclasses import dataclass

from oxidize_python.core.backends.factory import BackendConfig, create_compute_backend
from oxidize_python.core.model.inference import InferenceModel
from oxidize_python.core.model.loader import LoaderConfig, load_gguf_model_from_path
from oxidize_python.core.model.model import Model


@dataclass
class CachedModel:
    path: str
    model: Model
    backend: str = "cpu"
    warning: str = ""


class ModelCache:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._items: dict[str, CachedModel] = {}

    def _key(self, path: str, cfg: LoaderConfig) -> str:
        return f"{path}|{cfg.backend}|{cfg.ctx_size}|{cfg.n_gpu_layers}|{cfg.gpus}"

    def load(self, path: str, cfg: LoaderConfig) -> CachedModel:
        key = self._key(path, cfg)
        with self._lock:
            if key in self._items:
                return self._items[key]
        loaded = load_gguf_model_from_path(path, cfg)
        _, effective, warning = create_compute_backend(
            BackendConfig(name=cfg.backend, n_gpu_layers=cfg.n_gpu_layers, gpus=cfg.gpus)
        )
        entry = CachedModel(
            path=path,
            model=loaded,
            backend=effective or cfg.backend,
            warning=warning or "",
        )
        with self._lock:
            self._items[key] = entry
        return entry


default_model_cache = ModelCache()


def inference_from_cache(path: str, cfg: LoaderConfig) -> tuple[InferenceModel, CachedModel]:
    entry = default_model_cache.load(path, cfg)
    if not isinstance(entry.model, InferenceModel):
        raise TypeError(f"expected InferenceModel, got {type(entry.model)}")
    return entry.model, entry
